lib/archive/unpack_unix.go

```diff
@@ -131,71 +139,113 @@ func InvokeUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, r
 	return nil
 }
 
-// Unpack hooks into a binary present in the appliance vm called unpack in order to execute InvokeUnpack inside of a chroot. This method works identically to InvokeUnpack, except that it will not function in areas where the binary is not present at /bin/unpack
-func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+// OfflineUnpack wraps Unpack for usage in contexts without a childReaper.
+func OfflineUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+
+	var cmd *exec.Cmd
+	var err error
+	if cmd, err = unpack(op, tarStream, filter, root, applianceBinaryPath); err != nil {
+		return err
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func OnlineUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) (*exec.Cmd, error) {
+	return unpack(op, tarStream, filter, root, containerBinaryPath)
+}
+
+func streamCopy(op trace.Operation, stdin io.WriteCloser, tarStream io.Reader) {
+	// if we're passed a stream that doesn't cast to a tar.Reader copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
+	var err error
+	tr, ok := tarStream.(*tar.Reader)
+	if !ok {
+		defer stdin.Close()
+		if _, err := io.Copy(stdin, tarStream); err != nil {
+			op.Errorf("Error copying tarStream: %s", err.Error())
+		}
+		return
+	}
+
+	tw := tar.NewWriter(stdin)
+	defer tw.Close()
+	var th *tar.Header
+	for {
+		th, err = tr.Next()
+		if err == io.EOF {
+			tw.Close()
+			return
+		}
+		if err != nil {
+			op.Errorf("error reading tar header %s", err)
+			return
+		}
+		op.Debugf("processing tar header: asset(%s), size(%d)", th.Name, th.Size)
+		err = tw.WriteHeader(th)
+		if err != nil {
+			op.Errorf("error writing tar header %s", err)
+			return
+		}
+		switch th.Typeflag {
+		case tar.TypeReg:
```

Comment:
What about `TypeRegA`?

Reply:
Discussed on Slack, appears to be deprecated.

Reply:
Covering the case should not hurt. and might cover an edge case. I am curious if you have tried this by piping a tar file with directory entries and such directly into a docker cp? so `cat tar_with_dirs_and_files.tar | docker cp - <cid>:/some/path` since this looks to only be writing the TypeReg. It is true that the TypeDir ad such should not have a body that needs to be written so this is likely good. I just want to make sure if possible.

Reply:
update: worked with @gigawhitlocks on this out of band. I am convinced that things are working as they should.

Reply:
@gigawhitlocks did the slack discussion cover @hmahmood 's point about just copying data regardless of type and leaning on `io.Copy` to return `0, io.EOF`?

Reply:
@matthewavery was under the impression that the case that @hmahmood was concerned about could be exercised by copying a tarball that contains a directory w/ files in it. I copied him this shell transcript of myself exercising that case (it is also exercised in the tests, although the tests do `docker cp foo.tar.gz container:/dest` instead of `cat foo.tar.gz | docker cp - container:/dest`

```
✘ ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" volume create vol1                                                                                                           [14:20:55]
vol1

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" run --name on -v vol1:/vol1 -itd busybox                                                                                     [14:21:10]
a4d8aadad482c07edc21d544c6aedc39d00b4cfe119e0e494bb05e4d9b02155e

>>> elapsed time 46s                                                                                                                                                                                                                                                                                 

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯                                                                                                                                                                                                                                                                                        [14:21:57]
>>> elapsed time 47s                                                                                                                                                                                                                                                                                 

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" create --name off -v vol1:/vol1 -it busybox                                                                                  [14:22:19]
71aa580a25f6911ee5220e86d463cd436a2936fee36d854bd896f4ffe6ff6d0c
>>> elapsed time 18s                                                                                                                                                                                                                                                                                 

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ cat good.tar.gz | docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" cp - off:/vol1                                                                                             [14:23:00]
>>> elapsed time 33s                                                                                                                                                                                                                                                                                 

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" exec on ls /vol1                                                                                                             [14:23:56]
good

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" exec on ls /vol1/good                                                                                                        [14:24:01]
foo
good.tar.gz
hostd.log
secrets.yml
service-swagger-gen.log
swagger-gen.log
vic-machine.log

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" exec on ls /vol1/good/foo                                                                                                    [14:24:10]
bar

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ tar xvf good.tar.gz                                                                                                                                                                                                                                                                    [14:24:21]
good/
good/good.tar.gz
good/service-swagger-gen.log
good/secrets.yml
good/foo/
good/foo/bar
good/vic-machine.log
good/hostd.log
good/swagger-gen.log

 ian@trantor  ~/g/s/g/v/vic   pull-tar-bug ⁝ S ? 
 ❯ cat evil-image/archive.tar.gz | docker -H sc-rdops-vm07-dhcp-206-183.eng.vmware.com:2376 --tlsverify --tlscacert="ian-vch/ca.pem" --tlscert="ian-vch/cert.pem" --tlskey="ian-vch/key.pem" cp - off:/vol1                                                                               [14:24:43]
Error response from daemon: Server error from portlayer: error writing files to device vol1
```

Reply:
@hmahmood does this satisfy your concern about `TypeRegA`?